### PR TITLE
Best plane fit from points

### PR DESCRIPTION
**Pull request by f-frhs**
I added `Plane.CreateFittedPlaneFrom(points)` method with some unit tests for it.

I would like to discuss the following topic:
- [ ] Were testcases I added sufficient?
- implementation
  - [x] which is nice to evaluate?
    a. (z)axis-aligned distance
    b. orthogonal distance

### CODE REVIEW

File: src/Spatial/Euclidean/Plane.cs [hunk 1]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Plane"/> struct.
+        /// Constructs a Plane from the given points.
+        /// </summary>
+        /// <param name="points">The points</param>
+        /// <returns>Fitted <see cref="Plane"/></returns>
+        /// <remarks>this method uses SVD to fit the plane.</remarks>
+        public static Plane CreateFittedPlaneFrom(IEnumerable<Point3D> points)
+        {
+            var c = Point3D.Centroid(points);
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:89]: I would use more descriptive variable names, like `throughPoint` instead of just `c`.

File: src/Spatial/Euclidean/Plane.cs [hunk 2]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
+
+            var A = CreateMatrix.DenseOfRowVectors(
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:91]: This is also called the [scatter matrix](https://math.stackexchange.com/questions/3501135/fitting-a-plane-to-points-using-svd)

File: src/Spatial/Euclidean/Plane.cs [hunk 3]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
+                points.Select(p => p - c)
+                    .Select(p => CreateVector.DenseOfArray(new double[] { p.X, p.Y, p.Z })));
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:93]: What I would do in this case, is to first allocate the necessary matrix space (`new DenseMatrix(rows, columns)`), and then write a single for-each to populate each row of it. This way we could easily parallelize the loop in case we need it, if for example we have a LOT of points.

File: src/Spatial/Euclidean/Plane.cs [hunk 4]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
+
+            var svd = A.Svd(true);
+            var matV = svd.VT.Transpose();
+            var theIndex = svd.S.Count-1; // in this case, theIndex = 2.
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:97]: I think you can entirely skip this variable and use `var normal = UnitVector3D.OfVector(matV.Column(matV.Cols - 1));`. This better captures semantically the intention, that we need the last eigenvector of V.

File: src/Spatial/Euclidean/Plane.cs [hunk 5]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
+            var normal = UnitVector3D.OfVector(matV.Column(theIndex));
+
+            var result = new Plane(normal, c);
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:100]: `var bestFit = new Plane(normal, c);`

File: src/Spatial/Euclidean/Plane.cs [hunk 6]
```diff
@@ -74,6 +77,30 @@ public Plane(Point3D rootPoint, UnitVector3D normal)
...
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Plane"/> struct.
+        /// Constructs a Plane from the given points.
+        /// </summary>
+        /// <param name="points">The points</param>
+        /// <returns>Fitted <see cref="Plane"/></returns>
+        /// <remarks>this method uses SVD to fit the plane.</remarks>
+        public static Plane CreateFittedPlaneFrom(IEnumerable<Point3D> points)
```
Comment (jkalias) [inline src/Spatial/Euclidean/Plane.cs:87]: I think we need to locally transform the `IEnumerable<Point3D>` to a `List<Point3D>`. Then we wouldn't traverse the collection multiple times.

I also think that a better name would just be `BestFit`

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 7]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
             AssertGeometry.AreEqual(expected, projectedPoint, eps);
         }
 
+        [TestCase(ZeroPoint, X, Y)] //all 3 points are on the plane z=0. normal vector is +ez.
+        [TestCase(ZeroPoint, Y, Z)] //all 3 points are on the plane x=0. normal vector is +ex.
+        [TestCase(ZeroPoint, Z, X)] //all 3 points are on the plane y=0. normal vector is +ey.
+        public void CreateFittedPlaneFrom_PointsShouldBeOnTheFittedPlane(string sP1, string sP2, string sP3)
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:91]: I would also add another test case, with 3 general points, not just the ones which lie on the x, y and z axes.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 8]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+            var ps = new[]
+            {
+                Point3D.Parse(sP1),
+                Point3D.Parse(sP2),
+                Point3D.Parse(sP3),
+            }.ToArray();
+            var aPlane = Plane.CreateFittedPlaneFrom(ps);
+
+            var tolerance = 1e-16;
+            Assert.That(aPlane.SignedDistanceTo(ps[0]), Is.EqualTo(0).Within(tolerance), "ps[0]");
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:102]: `AssertGeometry.AreEqual(expectedPlane, actualPlane);`

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 9]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+            Assert.That(aPlane.SignedDistanceTo(ps[2]), Is.EqualTo(0).Within(tolerance), "ps[2]");
+        }
+
+        [TestCase(ZeroPoint, X, Y, XY, Z, "0,0,+1")] //all 4 points fall on the plane z=+1
+        [TestCase(ZeroPoint, X, Y, XY, Z, "0,0,-1")] //all 4 points fall on the plane z=-1
+        [TestCase(ZeroPoint, Y, Z, YZ, X, "+1,0,0")] //all 4 points fall on the plane x=+1
+        [TestCase(ZeroPoint, Y, Z, YZ, X, "-1,0,0")] //all 4 points fall on the plane x=-1
+        [TestCase(ZeroPoint, Z, X, ZX, Y, "0,+1,0")] //all 4 points fall on the plane y=+1
+        [TestCase(ZeroPoint, Z, X, ZX, Y, "0,-1,0")] //all 4 points fall on the plane y=-1
+        public void CreateFittedPlaneFrom_With4PointsOnAxisAlignedPlane(string sV1, string sV2, string sV3, string sV4, string sENormal, string sRootPoint)
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:113]: It's not really clear to me, what this test is supposed to guarantee.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 10]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+            }.Select(deltaV => rootPoint + deltaV);
+            var actual = Plane.CreateFittedPlaneFrom(ps);
+
+            var eNormal = UnitVector3D.Parse(sENormal);
+            var eRootPoint = Point3D.Parse(sRootPoint);
+            var expected = new Plane(eNormal, eRootPoint);
+            Assert.That(actual, Is.EqualTo(expected));
+        }
+
+        [TestCase("1,1,9;1,2,14;1,3,20;2,1,11;2,2,17;2,3,23;3,1,15;3,2,20;3,3,26", -21.7240278973, -41.0640090729, 7.3269978417, 1)]
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:131]: We need more tests, which cover what happens when we have less than 3 points (2, 1, or none).

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 11]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+
+            //for debugging purpose
+            TestContext.WriteLine($"{nameof(actual)}={actual}");
+            TestContext.WriteLine($"{nameof(expected)}={expected}");
+
+            var tolerance = 6e-3; // for this example
+            Assert.That(actual.A, Is.EqualTo(expected.A).Within(tolerance), "A");
+            Assert.That(actual.B, Is.EqualTo(expected.B).Within(tolerance), "B");
+            Assert.That(actual.C, Is.EqualTo(expected.C).Within(tolerance), "C");
+            Assert.That(actual.D, Is.EqualTo(expected.D).Within(tolerance), "D");
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:156]: `AssertGeometry.AreEqual(expectedPlane, actualPlane);`

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 12]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+            // eq2:-Ax - By - Cz - D = 0
+            var normal = new Vector3D(eA, eB, eC);
+            var eNormalizedOffset = eOffset / normal.Length;
+            var expected = new Plane(normal.Negate().Normalize(), eNormalizedOffset);
+
+            //for debugging purpose
+            TestContext.WriteLine($"{nameof(actual)}={actual}");
+            TestContext.WriteLine($"{nameof(expected)}={expected}");
+
+            var tolerance = 6e-3; // for this example
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:152]: This seems like a magic number. Perhaps the values from the original example are not as exact as we want them to be. I would adjust the values in our example, so that the default tolerance would pass the test.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs [hunk 13]
```diff
@@ -80,6 +85,77 @@ public void ProjectPointOn(string ps, string rootPoint, string unitVector, strin
...
+
+            //normalize (eA, eB, eC and eD).
+            //both 2 following equations mean the same plane.
+            // eq1: Ax + By + Cz + D = 0
+            // eq2:-Ax - By - Cz - D = 0
+            var normal = new Vector3D(eA, eB, eC);
+            var eNormalizedOffset = eOffset / normal.Length;
+            var expected = new Plane(normal.Negate().Normalize(), eNormalizedOffset);
+
+            //for debugging purpose
```
Comment (jkalias) [inline src/Spatial.Tests/Euclidean/PlaneTest.cs:148]: We don't need debugging for tests

**Review by jkalias — CHANGES_REQUESTED** [review]
Thank you for this PR, it seems interesting. Let me know if my comments are not clear enough.

**Comment by jkalias** [comment]
>       * [ ]  which is nice to evaluate?
>         a. (z)axis-aligned distance
>         b. orthogonal distance

I think your implementation is correct, we should evaluate the best fit based on the orthogonal euclidean distance.

File: src/Spatial/Euclidean/Plane.cs (hunk 1 above)
Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:89]: will do, thanks.

File: src/Spatial/Euclidean/Plane.cs (hunk 2 above)
Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:91]: If I understood your comment correctly, it seems to me that the [scatter matrix](https://math.stackexchange.com/questions/3501135/fitting-a-plane-to-points-using-svd) means a different matrix from the matrix `A` written in the code.

To say that these 2 matrices are different, I checked their sizes(=rows count and columns count).
The [scatter matrix](https://math.stackexchange.com/questions/3501135/fitting-a-plane-to-points-using-svd) is an 3x3 matrix, 
and the matrix `A` is an nx3 matrix,
where $n$ is the count of points.

I agree with you to give variable a descriptive name.
Do you have any ideas?

File: src/Spatial/Euclidean/Plane.cs (hunk 3 above)
Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:93]: will do, thanks.
I think that the parallelization is out of scope for me.
Thanks again for your point of view.

Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:93]: Does this fix follow the policy of Math.NET Spatial?
I'm afraid the case where you maintainers respect RAII([Resource acquisition is initialization](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)) principle and I would break it.

File: src/Spatial/Euclidean/Plane.cs (hunk 5 above)
Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:100]: I'm afraid that I didn't define any properties the *best* Fit plane should have.
Can I use `fittedPlane` for this?

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 9 above)
Comment (f-frhs) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:113]: I totally agree with you.
I think I should remove this test.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 10 above)
Comment (f-frhs) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:131]: `Plane.CreateFittedPlaneFrom(points)` is not for the testcases with insufficient points.
`Plane.CreateFittedPlaneFrom(points)` is for over-determined system.
Instead, `Plane.Validate(points)` will do for this purpose.

It goes a detailed implementation, `Plane.Validate(points)` will be something like:
```csharp
public static void Validate(IEnumerable<Point3D> points)
{
    //      [p0x, p0y, p0z]
    //      [p1x, p1y, p1z]
    // mat= [p2x, p2y, p2z]
    //      [p3x, p3y, p3z]
    //      [p4x, p4y, p4z]
    //      [..., ..., ...]

    //this expression will be fixed, according to the discussion
    var mat = CreateMatrix.DenseOfRowVectors(
                points.Select(p => CreateVector.DenseOfArray(new double[] { p.X, p.Y, p.Z })));

    var rankOfMat = mat.Rank();  // the library uses SVD.
    if(rankOfMat <= 0) // DoF=0
    {
        throws new Exception("The same point?");
    }
    if(rankOfMat <= 1) // DoF=1
    {
        throws new Exception("Are they collinear?");
    }

    // do nothing intentionally
    // because there are no problem, we can span 2d-vector-space with given points
}
```



If few (2 or less) points are given, the method `Plane.TryFromPonits()` should return `false`. It is similar to `int.TryParse(string)`'s behavior.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 11 above)
Comment (f-frhs) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:156]: True. I wish I could have.
I need time to adjust `AssertGeometry.AreEqual(PlaneA, PlaneB)` to pass this test
where
```csharp
PlaneA.Normal == -PlaneB.Normal;
PlaneA.RootPoint == PlaneB.RootPoint;
```

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 12 above)
Comment (f-frhs) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:152]: ok. thank you for your cooperation.

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 13 above)
Comment (f-frhs) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:148]: thanks for a comment.
I'll remove them.

File: src/Spatial/Euclidean/Plane.cs (hunk 5 above)
Comment (f-frhs) [reply src/Spatial/Euclidean/Plane.cs:100]: > we should evaluate the best fit based on the orthogonal euclidean distance.

Sorry, I couldn't read all your comments at that moment. my fault.
You had already defined "the best fit" for this problem.
Now, we can use `var bestFit = new Plane(normal, c);`

File: src/Spatial/Euclidean/Plane.cs (hunk 2 above)
Comment (jkalias) [reply src/Spatial/Euclidean/Plane.cs:91]: My apologies, you are right. The scatter matrix (`M` in the paper) would be `A^T * A` in your notation. Perhaps we can call this `relativePointMatrix`, since this is exactly what it is.

File: src/Spatial/Euclidean/Plane.cs (hunk 3 above)
Comment (jkalias) [reply src/Spatial/Euclidean/Plane.cs:93]: > I think that the parallelization is out of scope for me.

Sure, I wasn't expecting this to be part of this PR. I don't want to prematurely optimize things in any case.

Comment (jkalias) [reply src/Spatial/Euclidean/Plane.cs:93]: > Does this fix follow the policy of Math.NET Spatial?
> I'm afraid the case where you maintainers respect RAII([Resource acquisition is initialization](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)) policy and I would break it.

I am not sure I understand your question/concern here, can you explain it again please?

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 10 above)
Comment (jkalias) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:131]: I think this is over-engineering the problem. We don't need a `Plane.Validate` method, we only need to check in this "best fit method" that before we dive in and calculate the SVD of A, its row count is more than or equal to 3. Something like `if (A.Rows < 3) { throw ...}` (the syntax might not be 100% perfect)

File: src/Spatial.Tests/Euclidean/PlaneTest.cs (hunk 11 above)
Comment (jkalias) [reply src/Spatial.Tests/Euclidean/PlaneTest.cs:156]: This is another problem I spotted in the `Plane` implementation after I took over the maintenance of this repo; the plane is not defined in its [Hesse normal form](https://en.wikipedia.org/wiki/Hesse_normal_form), which I find problematic.

Then the normal direction is uniquely defined and fixed, for any plane. In addition the equality operator should be adjusted, so that we don't check equality of the root points, but that something like `other.Conatins(rootPoint)`, where we check that the root point is contained in the other plane. There is nothing special in the choice of a root point for a plane, any point lying on it is equally valid.

I wanted to open another PR and fix this (breaking) change for the next (v0.8.0) release. Maybe you want to wait with this PR after I am finished (?)

**Review by jkalias — APPROVED** [review]

**Comment by jkalias** [comment]
Fixed accuracy test in #218
